### ISSUE TITLE
Incorrect type inference

### ISSUE THREAD

**Issue by Yura52** [astral-sh/ty#2018]
### Summary

(Please feel free to edit the issue title as needed!)

In the following snippet, `foo` passes the type checks, but `bar` does not:

```py
def foo[T](a: None | T, b: None | T) -> T:
    if a is None:
        assert b is not None
    else:
        b = a
    return b


def bar[T](a: None | T, b: None | T) -> T:
    if a is None:
        assert b is not None
    else:
        if b is None:
            b = a
    return b
```

Namely, `ty check` says:

```
error[invalid-return-type]: Return type does not match returned value
  --> local/ty_test.py:15:12
   |
13 |         if b is None:
14 |             b = a
15 |     return b
   |            ^ expected `T@bar`, found `None | T@bar`
   |
  ::: local/ty_test.py:9:41
   |
 9 | def bar[T](a: None | T, b: None | T) -> T:
   |                                         - Expected `T@bar` because of return type
10 |     if a is None:
11 |         assert b is not None
   |
info: rule `invalid-return-type` is enabled by default
```

However, `b` always becomes `T` by the end of both functions, so the reported error seems to be false positive.

### Version

0.0.2

**Comment by oconnor663** [comment]
It looks like we're failing to unify the two branches? https://play.ty.dev/9830b46c-09df-4123-a317-4e058ccc86e5

```py
def bar(a: int | None, b: int | None):
    if a is None:
        assert b is not None
        reveal_type(b)  # int
    else:
        if b is None:
            b = a
        reveal_type(b)  # int
    reveal_type(b)      # int | None
```

**Comment by AlexWaygood** [comment]
It's possible this might be related to https://github.com/astral-sh/ty/issues/690

**Comment by carljm** [comment]
Yes this is #690 - thanks for the report!
